clippy_lints/src/unnecessary_indexing.rs

```diff
@@ -0,0 +1,159 @@
+use clippy_utils::ty::is_type_diagnostic_item;
+use clippy_utils::visitors::for_each_expr;
+use rustc_ast::{BorrowKind, LitKind, Mutability};
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, Local, Node, UnOp};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for the use of `seq.is_empty()` in an if-conditional where `seq` is a slice, array, or Vec,
+    /// and in which the first element of the sequence is indexed.
+    ///
+    /// ### Why is this bad?
+    /// This code is unnecessarily complicated and can instead be simplified to the use of an
+    /// if..let expression which accessed the first element of the sequence.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let a: &[i32] = &[1];
+    /// if !a.is_empty() {
+    ///    let b = a[0];
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let a: &[i32] = &[1];
+    /// if let Some(b) = a.first() {
+    ///
+    /// }
+    /// ```
+    #[clippy::version = "1.78.0"]
+    pub UNNECESSARY_INDEXING,
+    complexity,
+    "unnecessary use of `seq.is_empty()` in a conditional when if..let is more appropriate"
+}
+
+declare_lint_pass!(UnnecessaryIndexing => [UNNECESSARY_INDEXING]);
+
+impl LateLintPass<'_> for UnnecessaryIndexing {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &'_ rustc_hir::Expr<'_>) {
+        if let Some(if_expr) = clippy_utils::higher::If::hir(expr)
+            // check for negation
+            && let ExprKind::Unary(UnOp::Not, unary_inner) = if_expr.cond.kind
+            // check for call of is_empty
+            && let ExprKind::MethodCall(method, conditional_receiver, _, _) = unary_inner.kind
+            && method.ident.as_str() == "is_empty"
+            && let expr_ty = cx.typeck_results().expr_ty(conditional_receiver).peel_refs()
+            && (expr_ty.is_array_slice() || expr_ty.is_array() || is_type_diagnostic_item(cx, expr_ty, sym::Vec))
+            && let ExprKind::Block(block, _) = if_expr.then.kind
+        {
+            // the receiver for the index operation
+            let mut index_receiver: Option<&Expr<'_>> = None;
+            // first local in the block - used as pattern for `Some(pat)`
+            let mut first_local: Option<&Local<'_>> = None;
+            // any other locals to be aware of, these are set to the value of `pat`
+            let mut extra_locals: Vec<&Local<'_>> = vec![];
+            // any other index expressions to replace with `pat` (or "element" if no local exists)
+            let mut extra_exprs: Vec<&Expr<'_>> = vec![];
+
+            for_each_expr(block.stmts, |x| {
+                if let ExprKind::Index(receiver, index, _) = x.kind
+                    && let ExprKind::Lit(lit) = index.kind
+                    && let LitKind::Int(val, _) = lit.node
+                    && eq_expr_value(cx, receiver, conditional_receiver)
```

Comment:
When `conditional_receiver` is a more complicated place such as a field expression we can run into tricky borrow situations e.g.

```rust
struct S {
    v: Vec<usize>,
}

impl S {
    fn method(&mut self) {}
}

fn main() {
    let mut s = S { v: vec![1, 2, 3] };

    if !s.v.is_empty() {
        let a = s.v[0];
        s.method();
        a;
    }
}
```

We can sidestep these by using [`path_to_local`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.path_to_local.html) initially to check that the receiver is a local rather than an arbitrary place, then [`path_to_local_id`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.path_to_local_id.html) instead of `eq_expr_value`